libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -0,0 +1,1074 @@
+					end := add(pos, <overallSize>)
+				}
+			)");
+			templ("functionName", functionName);
+
+			// TODO this can make use of CODECOPY for large strings once we have that in JULIA
+			size_t words = (value.size() + 31) / 32;
+			templ("overallSize", to_string(32 + words * 32));
+			templ("length", to_string(value.size()));
+			vector<map<string, string>> wordParams(words);
+			for (size_t i = 0; i < words; ++i)
+			{
+				wordParams[i]["offset"] = to_string(32 + i * 32);
+				wordParams[i]["wordValue"] = "0x" + h256(value.substr(32 * i, 32), h256::AlignLeft).hex();
+			}
+			templ("word", wordParams);
+			return templ.render();
+		}
+		else
+		{
+			solAssert(_to.category() == Type::Category::FixedBytes, "");
+			solAssert(value.size() <= 32, "");
+			Whiskers templ(R"(
+				function <functionName>(pos) {
+					mstore(pos, <wordValue>)
+				}
+			)");
+			templ("functionName", functionName);
+			templ("wordValue", "0x" + h256(value, h256::AlignLeft).hex());
+			return templ.render();
+		}
+	});
+}
+
+string ABIFunctions::abiEncodingFunctionFunctionType(
+	FunctionType const& _from,
+	Type const& _to,
+	bool _encodeAsLibraryTypes,
+	bool _compacted
+)
+{
+	solAssert(_from.kind() == FunctionType::Kind::External, "");
+	solAssert(_from == _to, "");
+
+	string functionName =
+		"abi_encode_" +
+		_from.identifier() +
+		"_to_" +
+		_to.identifier() +
+		(_compacted ? "_compacted" : "") +
+		(_encodeAsLibraryTypes ? "_lib" : "");
+
+	if (_compacted)
+	{
+		return createFunction(functionName, [&]() {
+			return Whiskers(R"(
+				function <functionName>(addr_and_function_id, pos) {
+					mstore(pos, <cleanExtFun>(addr_and_function_id))
+				}
+			)")
+			("functionName", functionName)
+			("cleanExtFun", cleanupCombinedExternalFunctionIdFunction())
+			.render();
+		});
+	}
+	else
+	{
+		return createFunction(functionName, [&]() {
+			return Whiskers(R"(
+				function <functionName>(addr, function_id, pos) {
+					mstore(pos, <combineExtFun>(addr, function_id))
+				}
+			)")
+			("functionName", functionName)
+			("combineExtFun", combineExternalFunctionIdFunction())
+			.render();
+		});
+	}
+}
+
+string ABIFunctions::combineExternalFunctionIdFunction()
+{
+	string functionName = "combine_external_function_id";
+	return createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(addr, selector) -> combined {
+				combined := <shl64>(or(<shl32>(addr), and(selector, 0xffffffff)))
+			}
+		)")
+		("functionName", functionName)
+		("shl32", shiftLeftFunction(32))
+		("shl64", shiftLeftFunction(64))
+		.render();
+	});
+}
+
+string ABIFunctions::copyToMemoryFunction(bool _fromCalldata)
+{
+	string functionName = "copy_" + string(_fromCalldata ? "calldata" : "memory") + "_to_memory";
+	return createFunction(functionName, [&]() {
+		if (_fromCalldata)
+		{
+			return Whiskers(R"(
+				function <functionName>(src, dst, length) {
+					calldatacopy(dst, src, length)
+					// clear end
+					mstore(add(dst, length), 0)
```

Comment:
It just removes potential garbage. The routine might write to more memory than it actually "returns" in the end. This is documented in the header.